src/lavinmq/mqtt/broker.cr

```diff
@@ -70,14 +70,20 @@ module LavinMQ
         @vhost.rm_connection(client)
       end
 
-      def publish(packet : MQTT::Publish)
+      def publish(user : Auth::User, packet : MQTT::Publish)
+        unless user.can_write?(@vhost.name, packet.topic)
+          raise LavinMQ::Exchange::AccessRefused.new(@exchange)
+        end
         @exchange.publish(packet)
       end
 
       def subscribe(client, topics)
         session = sessions.declare(client)
         headers = AMQP::Table.new({RETAIN_HEADER => true})
         topics.map do |tf|
+          unless client.user.can_read?(@vhost.name, tf.topic)
```

Comment:
This is wrong usage of `#can_read?`. For subscribe (bind) one must check both read on exchange and write on queue. But maybe this can work different with mqtt.

Reply:
Yeah, the session(queue) "belongs" to the client, and client A does not really have access to client B's session.. so i think it can work a bit different for MQTT?

but definitely check read for the exchange, and check whether the user can subscribe to topics at all.